src/librustc/traits/select.rs

```diff
@@ -3356,13 +3356,28 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                     predicate: predicate.value
                 }))
         }).collect();
+
         // We are performing deduplication here to avoid exponential blowups
         // (#38528) from happening, but the real cause of the duplication is
         // unknown. What we know is that the deduplication avoids exponential
-        // amount of predicates being propogated when processing deeply nested
+        // amount of predicates being propagated when processing deeply nested
         // types.
-        let mut seen = FxHashSet();
-        predicates.retain(|i| seen.insert(i.clone()));
+        //
+        // This code is hot enough that it's worth avoiding the allocation
+        // required for the FxHashSet when possible. Special-casing lengths 0,
+        // 1 and 2 covers roughly 75--80% of the cases.
```

Comment:
What about using sort + dedup? That would avoid the allocation, and also quickly exit on small numbers of elements.

Reply:
And `Predicate` is a large type, so moving the values around might be moderately expensive.

Reply:
Oh, I see, the set is used with `retain`. But you need to clone & move the values into a set anyway. If you want to avoid the cost of data movement, the set could be created using references into the original `Vec` instead. Except that doesBut you need to move the values into a set anyway. If you want to avoid the cost of data movement, the set could be created using references into the original `Vec`n't work because of the `retain`...